### ISSUE TITLE
Support Procfile commands with a Dockerfile-based deployment

### ISSUE THREAD

**Issue by pmclanahan** [dokku/dokku#1564]
I read through #1350 but I'm still a bit confused. I'd expect the following to work (dokku 4.1):

1. Have a Dockerfile deployment that works.
2. Add a Procfile that defines different commands for a `web` and a `worker` process type.
3. Deploy and have the `web` process scale to 1 using the command from the Procfile.
4. Scale the `worker` type to 1 using `ps:scale` and have another instance of the docker image started running the `worker` command from the Procfile.

Is this a thing that's possible? I'm unclear on why It'd be required to do this in the app/image itself since buildpack-based deploys do this exactly and are also docker images. If you do think it's possible w/o breaking some kind of other feature compatibility then I'm happy to work on it.

**Comment by michaelshobbs** [comment]
As I'm sure you've found, that scenario does not currently work. I consider the dockerfile deployment as *beta* and could use more rfc/feature dev for sure. I think the current implementation is still my original "first pass" and to that point all assumptions were that the dockerfile author would dictate what gets run via the `RUN` verb. Please feel free to amend this (or a new issue) with what you're thinking, how it'll play with the rest of the feature set and let's flesh out what the next version of dockerfile support looks like.

Some notes on your current scenario.

1. If we remove the check for a dockerfile, we'll need a better way to determine which deployment method a user wants
2. seems reasonable
3. what if there is no `web` process. i think we currently handle this scenario anyway
4. i definitely like getting the same `ps` functionality with both deployment methods

any other dockerfile users have thoughts or comments on this?

**Comment by pmclanahan** [comment]
Thanks for the thorough reply. That all seems quite reasonable and cool. 

> If we remove the check for a dockerfile, we'll need a better way to determine which deployment method a user wants

Wasn't suggesting that. I was more using that step as a baseline; a kind of "steps to reproduce". I can see how that sounded bad. The Dockerfile deploy works just fine now IMHO. I'm just wishing for the ability to alter the run command via the Procfile. Apologies for the confusing wording.

**Comment by pmclanahan** [comment]
> what if there is no web process. i think we currently handle this scenario anyway

I think so, yeah. I think it'd be a required thing if you wanted a Procfile. Could potentially fall back to the CMD from the Dockerfile as the default for `web` if it's not there, but it seems quite reasonable to require a valid Procfile to contain a `web` type.

**Comment by callahad** [comment]
FWIW, this seems to parallel the broader discussion in the Docker / Rkt / OCI world of what the primitive unit of containerized deployments should be. As imagined by Docker Inc., each container should generally only run one process, so if you have a `web` and `worker` process, you should have two different containers, rather than just two instances of the same container. Which then gets super awkward in other ways. Hence Docker Compose, Kubernetes Pods, etc.

TL;DR: I don't think there's any clear consensus on the outside world for how we should handle this use case, so we're probably pretty free to blaze our own trail.

**Comment by pmclanahan** [comment]
That's interesting. I had no idea there was a question there. In my head it seems obvious that image is a unit and instances of that unit (the container) could be running pretty much any imaginable permutation of executables and arguments contained therein. This certainly seems to be how Heroku handles it, but perhaps on the backend they have a separate container image for each proc type. I certainly don't claim any expertise here. I mainly just want to be able to use a Dockerfile as well as a Procfile so that I can easily and separately run and scale web and worker processes.

**Comment by mixxorz** [comment]
:+1:

Having the ability to use workers whilst using Dockerfile deploys is a big plus.

Having two instances of the "same" container is what happens when you scale your containers horizontally. I think it's pretty reasonable to assume that doing that is okay.

Running the same container with different commands is impossible. A container is created when you run a command against an image. When you run two different commands against the same image, you theoretically end up with two different containers.

In any case, I think what's being asked for here is pretty straightforward. If you have both a Dockerfile and Procfile, the Procfile determines how the containers are executed. If you don't, then it falls back to how it's executed now.

**Comment by mmerickel** [comment]
You might find this comment[1] useful. Most containers have an entry point that determines what to execute... for dokku to bypass this might be considered odd. Rather I think it's up to you to write a container that can start the appropriate service given the environment if that container happens to support multiple services. Perhaps someone could write a common entrypoint bash script people could use in their containers though. 

[1] https://github.com/dokku/dokku/pull/1596#issuecomment-170081612

**Comment by pmclanahan** [comment]
> When you run two different commands against the same image, you theoretically end up with two different containers.

Yes! I think most of the confusion here has been my fault for not fully appreciating this distinction in the terminology. What I want is the ability to run the same image with different commands specified in the Procfile; exactly as you said.

**Comment by mmerickel** [comment]
> What I want is the ability to run the same image with different commands specified in the Procfile; exactly as you said.

Just to clarify I think so far you are asking dokku to completely ignore the entrypoint/cmd defaults of the docker image and do its own thing with a Procfile. This is assuming that the commands in the procfile are even suitable to run as PID 1 and that the assumptions inside the container make any sense at all when ignoring the entrypoint/cmd. I think it's much better to provide a suitable runner/binary people can embed in their images (installed by the dockerfile) which can be set as the entrypoint/cmd defaults and inspect the dokku-provided environment to startup appropriate services.

**Comment by pmclanahan** [comment]
Yes, if there's a Procfile it can and should ignore the CMD from the image, exactly the same way it does when you run `dokku run`. This is the way Heroku and Deis work. I think the assumption is that the commands in Procfile are appropriate and if they're not then the deployment will fail just as it would with an invalid CMD in a Dockerfile. I don't see how we can have separate process types that can be independently scaled (like heroku) using the same app environment. I can already maintain separate apps using the same codebase but maintaining separate app configs is a bit tedious and unnecessary. And if we suddenly require some special Dokku-only entrypoint program to make this work then I'm not likely to use it as I like generic and portable images.

**Comment by josegonzalez** [comment]
What happens if I have an app that supports both dockerfile deploys and buildpack deploys with a Procfile that *isn't* what I want to run via docker?

**Comment by pmclanahan** [comment]
Is that a thing? That doesn't sound like a thing that dokku should concern itself with. I guess if you wanted to do that you'd have to add `Procfile` to `.dockerignore`?

**Comment by pmclanahan** [comment]
I apologize if I'm asking for something outlandish. I was just hoping for parity with Deis really. I commonly run apps that require a worker and having a Procfile and Dockerfile is how it works in other systems. Is there something in the environment that tells me the process type? As far as I can tell currently if I scale a `worker` process type using `dokku ps:scale worker=1` it will start up just another of the same container as the web process type. If I'm understanding correctly you are suggesting that dokku should not work like the other systems and just let you work out what to really run using the process type discoverable in the environment variables?

**Comment by josegonzalez** [comment]
It will set the `DYNO` variable to `worker.1` I think.

**Comment by pmclanahan** [comment]
Indeed it does. Just checked. This could certainly be a workaround, but I still think it'd be a more robust and appropriate solution to get parity with Deis here. But it's obviously up to y'all :)

**Comment by mmerickel** [comment]
>  Is there something in the environment that tells me the process type?

Sorry I assumed you would read my comment I linked from the other thread. The `DYNO` env var is set by dokku so your entrypoint can start the appropriate service. This is what I meant above by making the container dependent on its environment.

**Comment by josegonzalez** [comment]
My concern here is that now we're overriding docker stuff and coming up with something custom (whether Deis supports it or not isn't really my concern). If I locally have a Procfile to ease local development and then my dokku deploy ignores my `ENTRYPOINT`, I'd be rightfully pissed. I'd like to avoid that if necessary.

If anything, you could do something like this in your `entrypoint.sh`:

```shell
#!/usr/bin/env bash

main() {
  pushd /app > /dev/null
  wget --quiet -O /app/forego https://godist.herokuapp.com/projects/ddollar/forego/releases/current/linux-amd64/forego
  PROCESS=$(echo $DYNO | cut -d'.' -f 1)
  forego start $PROCESS
}

main "$@"
```

Which requires zero extra work on dokku - including workaround to disable procfile parsing on dockerfile deploys for instance - and is extendable by the developer.

**Comment by pmclanahan** [comment]
Dokku already seems to be ignoring Procfiles when a Dockerfile is present. It also seems to me like it'd be more consistent since build-pack deployments build docker images and the command is optionally specified in the Procfile. So then it makes sense to me that a Dockerfile deployment would just replace the build-pack docker image with your own but the rest should still work.

@mmerickel my apologies. I did read your comment but apparently focused more on the bit where you said the Procfile specified the types and failed to note the `DYNO` variable and its significance. Thanks for the note.

**Comment by josegonzalez** [comment]
@pmclanahan other than deis, is there another PaaS that automatically adds the above entrypoint?

**Comment by pmclanahan** [comment]
I don't think it adds an entrypoint, I think it just sets in the app config that the `$process_name` should start with `$command` and then uses `docker run $command` to make it all go. If you want that `$command` to be an argument to your `ENTRYPOINT` then that's on you. I believe this is also how Heroku's Docker support works but I'd have to double check.

**Comment by josegonzalez** [comment]
Can you confirm Heroku's docker support works this way (both with a real app and with a doc link)?

**Comment by pmclanahan** [comment]
I don't have an app, but could deploy one. Their Docker support isn't great TBH, because I think you have to use one of their supported base images, but it does seem to work this way with the Procfile based on [the docs](https://devcenter.heroku.com/articles/docker#release). I'll see if I can get an app running this way soon.

**Comment by pmclanahan** [comment]
@josegonzalez now that I have a decent workaround since you and @mmerickel have finally gotten into my head what you meant with the bootstrap script and the `DYNO` variable I think we can close this if you really don't like the idea. I agree that feature parity with Heroku and/or Deis isn't a necessary goal for the dokku project, but in this case it still makes the most sense in my head :)

Thanks for your time.

**Comment by josegonzalez** [comment]
I just want to follow what Heroku uses, as that will more or lease ease our pain. If they do that - and you can verify that to be the case - then I'm happy to build this out.

**Comment by pmclanahan** [comment]
Cool. I'm happy to give a shot at a PR for this as well. I'll try out Heroku soon and report back.

**Comment by mixxorz** [comment]
I just got it to work as well! Thanks guys!

For now, should we document this? At least until there's a more permanent solution?

**Comment by pmclanahan** [comment]
> For now, should we document this? At least until there's a more permanent solution?

:+1:

**Comment by pmclanahan** [comment]
I setup a test. It seems to be working the way I thought it did based on the docs, but check out my results and see what you think:

1. [repo with the files I used](https://github.com/pmclanahan/heroku-docker-test)
2. [console log of the heroku deployment](https://gist.github.com/pmclanahan/b2090bb317037e3805ce)

**Comment by josegonzalez** [comment]
Waiting on the PR :)

**Comment by pmclanahan** [comment]
Touché :) Thanks for looking at it again, I'll try to get a PR together soon.

**Comment by michaelshobbs** [comment]
At this current point in time, I am :-1: on the implementation in #1915. I'd like to at least have the conversation around supporting dockerfiles in a similar way to Heroku. See [here](https://devcenter.heroku.com/articles/docker) for more details.

**Comment by michaelshobbs** [comment]
Regarding https://github.com/dokku/dokku/pull/1915#issuecomment-184034605....

I hacked together dockerfile support without much thinking. (Mostly pulled from a fork of dokku , actually) So I'm wondering if we should follow their implementation more closely instead of blazing our own trail and all the pitfalls that entails. We're essentially wrapping docker's build functionality and slapping nginx in front of that container. Is that useful? Is there a better way to go about this?

**Comment by pmclanahan** [comment]
I believe #1915 is as close as we can get right now. If we were to fail to deploy a docker file simply because it had an `ENTRYPOINT` we'd have to manually check the image and reject it for now obvious reason. It doesn't work in heroku because they don't use docker at all in their stack. Their docker support is purely local and they examine your setup and approximate what you've done and send it to their API. You also can only use base images derived from their official language images or `heroku/cedar:14`. I don't think that's a path dokku should go down.

However, the basic ideas of using the spirit of the Procfile to allow a user to easily define process types and associate commands therewith is good I think. The same idea works well no matter how the image was created. The herokuish image happens to take care of the procfile parsing and management so dokku doesn't have to, but using the Procile in this way for the other kind of deployment adds consistency to the overall user experience of dokku and makes it behave much more like heroku in the doker situation.

**Comment by pmclanahan** [comment]
I'll also point out that I purposefully implemented #1915 by extracting the procfile from the docker image so that other kinds of deployment (pulling from docker hub e.g.) could be possible in the future without much change to how this part works.

**Comment by michaelshobbs** [comment]
@pmclanahan I should have started with a thank you for putting together #1915. I appreciate the minimalist approach of the change to be certain.

One thought I had about this was instead of allowing any random base image to be used, we could require herokuish be the starting point. That way we can always ensure access to the herokuish toolbelt. Thoughts there?

**Comment by pmclanahan** [comment]
I'm actually a big fan of having the ability to use any base image I want with dokku. If it were a requirement for using a Procfile I could possibly be convinced, but not a requirement for using dockerfile with dokku. I actually run some static sites and other things that don't require the bulk of a debian or ubuntu base image.

**Comment by pmclanahan** [comment]
Also thanks :)

**Comment by pmclanahan** [comment]
hmmm... and actually I think if you did use herokuish as your base then dokku would think it was a procfile build anyway... so it wouldn't enter the dockerfile deployment path of the code anyway. The `is_herokuish` function would return `true` at least.

**Comment by michaelshobbs** [comment]
That implementation would have to change, obviously. 😃

**Comment by michaelshobbs** [comment]
The point of requiring herokuish as the base image is to have a consistent way to manage the app inside the container. Meaning Procfile parsing, process execution, etc.

**Comment by pmclanahan** [comment]
> The point of requiring herokuish as the base image is to have a consistent way to manage the app inside the container. Meaning Procfile parsing, process execution, etc.

Understood. I just don't think that should be dokku's concern. It seems to me to be a much more simple docker container manager. In my head it's job is to provide a consistent environment for the execution of docker containers, and manage said containers and how many are running, but not much more. Why would we want to manage the app inside?

However, like I said in #1915, we can get away even cheaper by not supporting Procfile and just supplying a better env var for process type. `DYNO` sortof works but it includes the scale number and so isn't very straight forward. Then people could use that variable to do what they want, or use `foreman` or `forego` or whatever themselves. If we then did a good job with documenting this variable it would probably be good enough for a while and we could punt on the larger architectural questions.

**Comment by pmclanahan** [comment]
That said you do bring up an interesting situation that should likely be handled. What if someone today uses herokuish as their base for a dockerfile deployment. I'm not sure what would happen. We should probably make sure that works and is documented. And like I said I could get behind requiring herokuish base image if you want dokku to do the procfile stuff for you in lieu of #1915, but I would be very sad indeed if that meant the removal of generic dockerfile support.

**Comment by michaelshobbs** [comment]
This is definitely the conversation I was wanting to engage in. Fundamentally dokku started as a roll your own paas that mirrored heroku. Dockerfile support changed that trajectory but since it hasn't been widely used (until now) there weren't enough folks with a vested interest with which to discuss the feature set. 

To me the choice is, continue to be a Heroku clone of sorts and follow their feature support more closely or don't and build out a more generic docker container/image manager. The former means we should probably stick to requiring herokuish in all apps. The latter means we can do whatever the crap we want as a community with dockerfile support. 

To your implementation points. I like the flexibility of choosing your own process manager and the simpler approach if we go the route of adding additional ENV variables beyond `DYNO`. That being said, does sticking with a `Procfile` preclude using something like foreman or forego?

**Comment by pmclanahan** [comment]
It doesn't preclude it, but it does leave it as an exercise for the user.
What it basically means is that you don't have to use those process
managers and that you have more flexibility than heroku allows. The
question of what dokku wants to be is bigger than this issue I think, but
will inform this implementation for sure. I think I'd like to see support
for both tbh.

**Comment by josegonzalez** [comment]
@michaelshobbs the alternative is to support docker-compose. My initial experiment (I closed the PR) was not fruitful, and honestly I don't think that way is a path we can go down (at least with the tools we have at our disposal).

I think that allowing an easy way for developers to customize the command run from a single dockerfile is one of the last things we need for good dockerfile support, so if this is a way that a popular PaaS has done it, then I'm +1. We just need to document everything about it, have good tests, and ensure that everyone thinks this is a good way forward.

**Comment by michaelshobbs** [comment]
@josegonzalez I'm unclear as to what you are referring when you say: "...so if this is a way that a popular PaaS has done it..." Are you referring to simply supporting `Procfile` definitions within Dockerfile apps? To be clear, while Heroku does require a `Procfile`, they do not allow arbitrary images. This method is a way for us to support Dockerfile apps in a "Heroku-like" way and not exactly like Heroku does it. This is an important distinction as we will be taking on additional complexity in what dokku will be expected to support. 

For the record, I am fine with the decision either way. My intent is only to encourage a discussion about this as a community and be intentional about what we want to support going forward.

Also, :+1: to good documentation and tests! :smile:

**Comment by josegonzalez** [comment]
I don't see how allowing usage of Procfiles breaks custom images.

**Comment by michaelshobbs** [comment]
It doesn't. However, it does require us to execute the processes inside the container differently than a buildpack app. That's all. 

I also fully understand that's how our dockerfile support has worked since the beginning. My point to challenge this assumption in an effort to make sure we're being intentional about our path forward in support Dockerfile-based apps.

**Comment by josegonzalez** [comment]
How does it change? I'm not really understanding what you mean.

**Comment by michaelshobbs** [comment]
We will now have two methods of parsing a `Procfile`; `herokuish` and [`get_cmd_from_procfile()`](https://github.com/pmclanahan/dokku/blob/0eb34ef5cfd20ee89174dcd6d7adf5dc831108dc/plugins/common/functions#L281-L288). That's increased complexity. I'm fine with increasing complexity as long as we are intentional about it.

**Comment by josegonzalez** [comment]
Ah gotcha. Can we repurpose the procfile parsing from herokuish somehow?

**Comment by michaelshobbs** [comment]
Not without including the the `herokuish` binary. Hence my thought about requiring every Dockerfile app be based on the herokuish image. I suppose we could somehow inject the binary **shrug**?

**Comment by josegonzalez** [comment]
Hmmm can we pull the Procfile from the app, shove it into a temp herokuish container, and pull it that way?

**Comment by mmerickel** [comment]
Injecting the binary is likely too much of an assumption about the internal container. Also herokuish is currently an optional dependency on dokku.

**Comment by josegonzalez** [comment]
@mmerickel good catch. I guess parsing it in bash is our best bet then. @pmclanahan I'm okay with the added complexity, just needs solid tests and docs.

**Comment by michaelshobbs** [comment]
@mmerickel Of course this would need to change if we go down this path
@josegonzalez Honestly I'm looking for a simpler solution but I don't know that injecting `herokuish` is the right move either.

**Comment by josegonzalez** [comment]
@michaelshobbs I don't think herokuish is the right move. Maybe we can extract the procfile parsing into a standalone thing and package that with both dokku and herokuish?

**Comment by michaelshobbs** [comment]
That seems more complex TBH. The current implementation is probably fine for the time being.

**Comment by pmclanahan** [comment]
So there is a Procfile parser in bash called [shoreman](https://github.com/chrismytton/shoreman/). Doesn't look like we could really use it directly, but we could base our parsing on [theirs](https://github.com/chrismytton/shoreman/blob/master/shoreman.sh#L89-L96) and add tests.

**Comment by michaelshobbs** [comment]
@pmclanahan looks familiar :smile: 
https://github.com/gliderlabs/herokuish/blob/v0.3.8/include/procfile.bash#L11-L22

**Comment by pmclanahan** [comment]
Nice :) Also we could borrow some of that.

**Comment by alanjds** [comment]
Btw, is that accidentally closed by a referring commit?

**Comment by josegonzalez** [comment]
No, thats when it was merged into master.
